VerifyProfileEmail/handler.ts

```diff
@@ -0,0 +1,232 @@
+import { PatternString } from "italia-ts-commons/lib/strings";
+
+import { RequiredQueryParamMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_query_param";
+
+import { VerificationTokenEntity } from "io-functions-commons/dist/src/entities/verification_token";
+import { ProfileModel } from "io-functions-commons/dist/src/models/profile";
+import { retrieveTableEntity } from "io-functions-commons/dist/src/utils/azure_storage";
+import { ContextMiddleware } from "io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "io-functions-commons/dist/src/utils/request_middleware";
+import { ValidUrl } from "italia-ts-commons/lib/url";
+
+// A token must be in the following format:
+// ULID + ":" + crypto.randomBytes(12)
+const TokenQueryParam = PatternString("^[A-Za-z0-9]{26}:[A-Fa-f0-9]{24}$");
+type TokenQueryParam = t.TypeOf<typeof TokenQueryParam>;
+
+type IVerifyProfileEmailHandler = (
+  context: Context,
+  token: string
+) => Promise<IResponsePermanentRedirect | IResponseErrorValidation>;
+
+/**
+ * Returns a ValidUrl in case of success
+ */
+function verificationSuccessUrl(verificationCallbackUrl: string): ValidUrl {
+  return { href: `${verificationCallbackUrl}?result=success` };
+}
+
+/**
+ * Returns a ValidUrl with a error in case of failure
+ */
+function verificationFailureUrl(
+  verificationCallbackUrl: string,
+  error: string
+): ValidUrl {
+  return { href: `${verificationCallbackUrl}?result=failure&error=${error}` };
+}
+
+const TokenQueryParamMiddleware = RequiredQueryParamMiddleware(
+  "token",
+  TokenQueryParam
+);
+
+// Used in the callback
+export enum VerificationErrors {
+  GENERIC_ERROR = "GENERIC_ERROR",
+  INVALID_TOKEN = "INVALID_TOKEN",
+  TOKEN_EXPIRED = "TOKEN_EXPIRED"
+}
+
+export function VerifyProfileEmailHandler(
+  tableService: TableService,
+  verificationTokensTableName: string,
+  profileModel: ProfileModel,
+  verificationCallbackUrl: string
+): IVerifyProfileEmailHandler {
+  return async (context, token) => {
+    const logPrefix = `VerifyProfileEmail`;
+
+    // STEP 1: Find and verify verification token
+
+    // Get required data
+    const [tokenId, validator] = token.split(":");
+    const validatorHash = crypto
+      .createHash("sha256")
+      .update(validator)
+      .digest("hex");
+
+    // Retrive the entity from the table storage
+    const errorOrMaybeVerificationTokenEntity = await retrieveTableEntity(
+      VerificationTokenEntity,
+      tableService,
+      verificationTokensTableName,
+      tokenId,
+      validatorHash
+    );
+
+    if (isLeft(errorOrMaybeVerificationTokenEntity)) {
+      context.log.error(
+        `${logPrefix}|Error searching verification token|ERROR=${errorOrMaybeVerificationTokenEntity.value}`
+      );
+      return ResponsePermanentRedirect(
+        verificationFailureUrl(
+          verificationCallbackUrl,
+          VerificationErrors.GENERIC_ERROR
```

Comment:
is this a temporary error? perhaps we need to encode that information here, to be able to tell the user what to do next